src/server/coprocessor/endpoint.rs

```diff
@@ -486,14 +480,156 @@ fn get_chunk(chunks: &mut Vec<Chunk>) -> &mut Chunk {
     chunks.last_mut().unwrap()
 }
 
+#[inline]
+fn encode_row_from_cols(cols: &[ColumnInfo],
+                        h: i64,
+                        values: HashMap<i64, &[u8]>,
+                        data: &mut Vec<u8>)
+                        -> Result<RowMeta> {
+    let mut meta = RowMeta::new();
+    meta.set_handle(h);
+    for col in cols {
+        let col_id = col.get_column_id();
+        if let Some(v) = values.get(&col_id) {
+            data.extend_from_slice(v);
+            continue;
+        }
+        if col.get_pk_handle() {
+            box_try!(datum::encode_to(data, &[get_pk(col, h)], false));
+        } else if mysql::has_not_null_flag(col.get_flag() as u64) {
+            return Err(box_err!("column {} of {} is missing", col_id, h));
+        } else {
+            box_try!(datum::encode_to(data, &[Datum::Null], false));
+        }
+    }
+    meta.set_length(data.len() as i64);
+    Ok(meta)
+}
+
+#[derive(Debug, Clone)]
+struct SortRow {
+    key: Vec<Datum>,
+    meta: RowMeta,
```

Comment:
I don't think `SortRow` needs a `RowMeta` here. Only top N rows need to be encoded as `RowMeta`.